internal/controller/checkpointrestoreoperator_controller.go

```diff
@@ -214,90 +294,293 @@ func getCheckpointArchiveInformation(log logr.Logger, checkpointPath string) (*c
+		handleCheckpointsForLevel(log, details, "pod", Policy{
+			MaxCheckpoints:    int(*policy.MaxCheckpoints),
+			MaxCheckpointSize: *policy.MaxCheckpointSize * MB,
+			MaxTotalSize:      *policy.MaxTotalSize * MB,
+		})
+	} else if policy := findNamespacePolicy(details); policy != nil {
+		log.Info("Applying namespace-level policy", "namespace", details.namespace)
+		handleCheckpointsForLevel(log, details, "namespace", Policy{
+			MaxCheckpoints:    int(*policy.MaxCheckpoints),
+			MaxCheckpointSize: *policy.MaxCheckpointSize * MB,
+			MaxTotalSize:      *policy.MaxTotalSize * MB,
+		})
+	} else {
+		// Apply global policies if no specific policy found
+		log.Info("Applying global container-level policy", "container", details.container)
+		handleCheckpointsForLevel(log, details, "container", Policy{
+			MaxCheckpoints:    maxCheckpointsPerContainer,
+			MaxCheckpointSize: maxCheckpointSize * MB,
+			MaxTotalSize:      maxTotalSizePerContainer * MB,
+		})
+
+		log.Info("Applying global pod-level policy", "pod", details.pod)
+		handleCheckpointsForLevel(log, details, "pod", Policy{
+			MaxCheckpoints:    maxCheckpointsPerPod,
+			MaxCheckpointSize: maxCheckpointSize * MB,
+			MaxTotalSize:      maxTotalSizePerPod * MB,
+		})
+
+		log.Info("Applying global namespace-level policy", "namespace", details.namespace)
+		handleCheckpointsForLevel(log, details, "namespace", Policy{
+			MaxCheckpoints:    maxCheckpointsPerNamespace,
+			MaxCheckpointSize: maxCheckpointSize * MB,
+			MaxTotalSize:      maxTotalSizePerNamespace * MB,
+		})
+	}
+}
+
+func handleCheckpointsForLevel(log logr.Logger, details *checkpointDetails, level string, policy Policy) {
+	if policy.MaxCheckpoints <= 0 && policy.MaxTotalSize <= 0 {
+		return
+	}
+
+	var globPattern string
+	switch level {
+	case "container":
+		globPattern = filepath.Join(
+			checkpointDirectory,
+			fmt.Sprintf(
+				"checkpoint-%s_%s-%s-[0-9][0-9][0-9][0-9]-[0-2][0-9]-[0-3][0-9]T[0-2][0-9]:[0-5][0-9]:[0-6][0-9]*.tar",
+				details.pod,
+				details.namespace,
+				details.container,
+			),
+		)
+	case "pod":
+		globPattern = filepath.Join(
+			checkpointDirectory,
+			fmt.Sprintf(
+				"checkpoint-%s_%s-*-*.tar",
+				details.pod,
+				details.namespace,
+			),
+		)
+	case "namespace":
+		globPattern = filepath.Join(
+			checkpointDirectory,
+			fmt.Sprintf(
+				"checkpoint-*_%s-*-*.tar",
+				details.namespace,
+			),
+		)
+	}
+
+	log.Info("Looking for checkpoint archives", "pattern", globPattern)
 	checkpointArchives, err := filepath.Glob(globPattern)
 	if err != nil {
-		log.Error(
-			err,
-			"error looking for checkpoint archives with",
-			"pattern",
-			globPattern,
-		)
+		log.Error(err, "error looking for checkpoint archives", "pattern", globPattern)
 		return
 	}
 
-	checkpointArchivesCounter := len(checkpointArchives)
-	if checkpointArchivesCounter <= maxCheckpointsPerContainer {
-		// Nothing to clean up, return early.
-		return
+	var filteredArchives []string
```

Comment:
@adrianreber @rst0git @snprajwal I need some assist on this approach .. currently I am filtering the archives if they have more specific policy defined, but it seems inefficient, what do you think on this??

Reply:
I think the approach is alright, since it'll take O(n) time in the worst case for n archives. If you want to make the logic a little more intuitive, you can reduce it to just two conditional blocks:
```go
if level != "container"  && findContainerPolicy(archiveDetails) != nil {
	continue
} else if level != "pod"  && findPodPolicy(archiveDetails) != nil {
	continue
}
```